crates/wasmparser/src/validator/component.rs

```diff
@@ -670,9 +677,23 @@ impl ComponentState {
 
         let state = components.pop().unwrap();
 
-        Ok(InstanceType {
+        // Push a component type for the instance
+        let component_ty = TypeDef::Component(ComponentType {
```

Comment:
I think it would be better to have a new enum of `TypeDef` which is something like `TypeDef::Instance` or something like that. Otherwise I'm afraid in the future that a function will be implemented that iterates over `types` or somehow relies on the information in the `types` array and this model could mistakenly classify an instance as a component. Overall I'd prefer to future-proof this and add a separate variant for module instance and component instance types which have a pointer (as you've implemented right now) back to the component it was instantiated from.

Reply:
Rather than expanding `TypeDef`, I added a "kind" enum to `ModuleInstanceType` and `InstanceType` that better represents how the instance type ended up in the type list.

Removed having to insert a component / module type to reference.